src/uct/base/uct_iface.c

```diff
@@ -714,31 +714,42 @@ uct_keepalive_info_t* uct_ep_keepalive_create(pid_t pid)
                                    &start_time);
     if (status != UCS_OK) {
         ucs_error("failed to get process start time");
-        ucs_free(ka);
-        return NULL;
+        goto err_free_ka;
     }
 
     ka->start_time = start_time;
 
     return ka;
+
+err_free_ka:
+    ucs_free(ka);
+err:
+    return NULL;
 }
 
 ucs_status_t
-uct_ep_keepalive_check(uct_ep_h tl_ep, uct_keepalive_info_t *ka, unsigned flags,
-                       uct_completion_t *comp)
+uct_ep_keepalive_check(uct_ep_h tl_ep, uct_keepalive_info_t **ka, pid_t pid,
+                       unsigned flags, uct_completion_t *comp)
 {
     ucs_status_t status;
     ucs_time_t create_time;
 
     UCT_EP_KEEPALIVE_CHECK_PARAM(flags, comp);
 
-    ucs_assert(ka != NULL);
-
-    status = ucs_sys_get_file_time(ka->proc, UCS_SYS_FILE_TIME_CTIME,
-                                   &create_time);
-    if (ucs_unlikely((status != UCS_OK) || (ka->start_time != create_time))) {
-        return uct_iface_handle_ep_err(tl_ep->iface, tl_ep,
-                                       UCS_ERR_ENDPOINT_TIMEOUT);
+    if (ucs_unlikely(*ka == NULL)) {
+        *ka = uct_ep_keepalive_create(pid);
+        if (*ka == NULL) {
+            return uct_iface_handle_ep_err(tl_ep->iface, tl_ep,
+                                           UCS_ERR_NO_MEMORY);
+        }
```

Comment:
what if remote pid is already dead? need to check it

Reply:
`uct_ep_keepalive_create` would fail then

Reply:
@yosefe do you agree with that?

Reply:
i see, so why is the error UCS_ERR_NO_MEMORY?

Reply:
fixed